rpcserver.go

```diff
@@ -6730,6 +6733,347 @@ func (r *rpcServer) EncodeCustomRecords(_ context.Context,
 	}
 }
 
+// SendPayment is a wrapper around lnd's routerrpc.SendPaymentV2 RPC method
+// with asset specific parameters. It allows RPC users to send asset keysend
+// payments (direct payments) or payments to an invoice with a specified asset
+// amount.
+func (r *rpcServer) SendPayment(req *tchrpc.SendPaymentRequest,
+	stream tchrpc.TaprootAssetChannels_SendPaymentServer) error {
+
+	if req.PaymentRequest == nil {
+		return fmt.Errorf("payment request must be specified")
+	}
+	pReq := req.PaymentRequest
+	ctx := stream.Context()
+
+	// Do some preliminary checks on the asset ID and make sure we have any
+	// balance for that asset.
+	if len(req.AssetId) != sha256.Size {
+		return fmt.Errorf("asset ID must be 32 bytes")
+	}
+	var assetID asset.ID
+	copy(assetID[:], req.AssetId)
+
+	// Now that we know we have at least _some_ asset balance, we'll figure
+	// out what kind of payment this is, so we can determine _how many_
+	// asset units we need.
+	destRecords := pReq.DestCustomRecords
+	_, isKeysend := destRecords[record.KeySendType]
+	firstHopRecords := pReq.FirstHopCustomRecords
+
+	switch {
+	// Both payment request and keysend is set, which isn't a supported
+	// combination.
+	case pReq.PaymentRequest != "" && isKeysend:
+		return fmt.Errorf("payment request and keysend custom " +
+			"records cannot be set at the same time")
+
+	// There are custom records for the first hop set, which means RFQ
+	// negotiation has already happened (or this is a keysend payment and
+	// the correct asset amount is already encoded). So we don't need to do
+	// anything special and can just forward the payment to lnd.
+	case len(firstHopRecords) > 0:
+		// Continue below.
+
+	// The request wants to pay a specific invoice.
+	case pReq.PaymentRequest != "":
+		invoice, err := zpay32.Decode(
+			pReq.PaymentRequest, r.cfg.Lnd.ChainParams,
+		)
+		if err != nil {
+			return fmt.Errorf("error decoding payment request: %w",
+				err)
+		}
+
+		// The peer public key is optional if there is only a single
+		// asset channel.
+		var peerPubKey *route.Vertex
+		if len(req.PeerPubkey) > 0 {
+			parsedKey, err := route.NewVertexFromBytes(
+				req.PeerPubkey,
+			)
+			if err != nil {
+				return fmt.Errorf("error parsing peer pubkey: "+
+					"%w", err)
+			}
+
+			peerPubKey = &parsedKey
+		}
+
+		// We can now query the asset channels we have.
+		assetChan, err := r.rfqChannel(ctx, assetID, peerPubKey)
+		if err != nil {
+			return fmt.Errorf("error finding asset channel to "+
+				"use: %w", err)
+		}
+
+		// Even if the user didn't specify the peer public key before,
+		// we definitely know it now. So let's make sure it's always
+		// set.
+		peerPubKey = &assetChan.channelInfo.PubKeyBytes
+
+		// TODO(guggero): This should actually be the max BTC amount
+		// (invoice amount plus fee limit) in milli-satoshi, not the
+		// asset amount. Need to change the whole RFQ API to do that
+		// though.
+		maxAssetAmount := assetChan.assetInfo.LocalBalance
```

Comment:
I don't get the TODO
Isn't this the asset local balance?

Reply:
It is currently, but will the threshold will be lowered to match the payment rather than just the entire channel outgoing balance.

Reply:
We set it to the asset balance because the unit is wrong (right now it's the maximum amount in _assets_, but the user doesn't know how many assets they need to send, that's the question the RFQ call should answer in the first place). We'll do some refactor and eventually this should be the maximum amount the user is willing to pay in **satoshis** (invoice amount plus the user-defined maximum routing fee they're willing to pay).